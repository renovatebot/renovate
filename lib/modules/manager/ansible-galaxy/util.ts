import { regEx } from '../../../util/regex';

export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx = regEx(
  /^\s+(?<packageName>[\w.]+):\s*["'](?<version>.+)["']\s*/
);
export const nameMatchRegex = regEx(
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/
);
