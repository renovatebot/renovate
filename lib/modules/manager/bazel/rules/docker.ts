import { z } from 'zod';
import { DockerDatasource } from '../../../datasource/docker';
import { id as dockerVersioning } from '../../../versioning/docker';
import type { PackageDependency } from '../../types';

export const dockerRules = ['container_pull'] as const;

export const DockerTarget = z
  .object({
    rule: z.enum(dockerRules),
    name: z.string(),
    tag: z.string(),
    digest: z.string(),
    repository: z.string(),
    registry: z.string(),
  })
  .transform(
    ({ rule, name, repository, tag, digest, registry }): PackageDependency => ({
      datasource: DockerDatasource.id,
      versioning: dockerVersioning,
      depType: rule,
      depName: name,
      packageName: repository,
      currentValue: tag,
      currentDigest: digest,
      registryUrls: [registry],
    })
  );
