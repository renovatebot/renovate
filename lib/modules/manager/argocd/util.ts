import { regEx } from '../../../util/regex';

export const keyValueExtractionRegex = regEx(
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/
);
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = regEx(/\s*apiVersion:\s*argoproj.io\/\s*/);
