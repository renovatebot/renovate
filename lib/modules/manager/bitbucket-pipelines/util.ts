import { regEx } from '../../../util/regex';
import { BitbucketTagsDatasource } from '../../datasource/bitbucket-tags';
import { getDep } from '../dockerfile/extract';
import type { PackageDependency } from '../types';

export const pipeRegex = regEx(`^\\s*-\\s?pipe:\\s*'?"?([^\\s'"]+)'?"?\\s*$`);
export const dockerImageRegex = regEx(
  `^\\s*-?\\s?image:\\s*'?"?([^\\s'"]+)'?"?\\s*$`
);
export const dockerImageObjectRegex = regEx('^(?<spaces>\\s*)image:\\s*$');

export function addDepAsBitbucketTag(
  deps: PackageDependency[],
  pipe: string
): void {
  const [depName, currentValue] = pipe.split(':');
  const dep: PackageDependency = {
    depName,
    currentValue,
    datasource: BitbucketTagsDatasource.id,
  };
  dep.depType = 'bitbucket-tags';
  deps.push(dep);
}

export function addDepAsDockerImage(
  deps: PackageDependency[],
  currentDockerImage: string
): void {
  const dep = getDep(currentDockerImage);
  dep.depType = 'docker';
  deps.push(dep);
}

export function addDepFromObject(
  deps: PackageDependency[],
  lines: string[],
  start: number,
  len: number,
  spaces: string
): number {
  const nameRegex = regEx(
    `^${spaces}\\s+name:\\s*['"]?(?<image>[^\\s'"]+)['"]?\\s*$`
  );
  const indentRegex = regEx(`^${spaces}\\s+`);

  for (let idx = start + 1; idx < len; idx++) {
    const line = lines[idx];

    if (!indentRegex.test(line)) {
      // malformed
      return idx;
    }

    const groups = nameRegex.exec(line)?.groups;
    if (groups) {
      const dep = getDep(groups.image);
      dep.depType = 'docker';
      deps.push(dep);
      return idx;
    }
  }

  // malformed
  return start;
}
