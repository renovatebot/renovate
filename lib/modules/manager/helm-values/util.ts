import { hasKey } from '../../../util/object';
import { regEx } from '../../../util/regex';
import type { HelmDockerImageDependency } from './types';

const parentKeyRe = regEx(/image$/i);

/**
 * Type guard to determine whether a given partial Helm values.yaml object potentially
 * defines a Helm Docker dependency.
 *
 * There is no exact standard of how Docker dependencies are defined in Helm
 * values.yaml files (as of February 26th 2021), this function defines a
 * heuristic based on the most commonly used format in the Helm charts:
 *
 * image:
 *   repository: 'something'
 *   tag: v1.0.0
 * image:
 *   repository: 'something'
 *   version: v1.0.0
 * renovateImage:
 *   repository: 'something'
 *   tag: v1.0.0
 */
export function matchesHelmValuesDockerHeuristic(
  parentKey: string,
  data: unknown
): data is HelmDockerImageDependency {
  return !!(
    parentKeyRe.test(parentKey) &&
    data &&
    typeof data === 'object' &&
    hasKey('repository', data) &&
    (hasKey('tag', data) || hasKey('version', data))
  );
}

export function matchesHelmValuesInlineImage(
  parentKey: string,
  data: unknown
): data is string {
  return !!(parentKeyRe.test(parentKey) && data && typeof data === 'string');
}
