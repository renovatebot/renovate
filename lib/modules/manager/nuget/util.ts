import upath from 'upath';
import { XmlDocument } from 'xmldoc';
import { logger } from '../../../logger';
import { findUpLocal, readLocalFile } from '../../../util/fs';
import { regEx } from '../../../util/regex';
import { defaultRegistryUrls } from '../../datasource/nuget';
import type { Registry } from './types';

export async function readFileAsXmlDocument(
  file: string
): Promise<XmlDocument | undefined> {
  try {
    // TODO #7154
    const doc = new XmlDocument((await readLocalFile(file, 'utf8'))!);
    // don't return empty documents
    return doc?.firstChild ? doc : undefined;
  } catch (err) {
    logger.debug({ err, file }, `failed to parse file as XML document`);
    return undefined;
  }
}

const defaultRegistries = defaultRegistryUrls.map(
  (registryUrl) => ({ url: registryUrl } as Registry)
);

export function getDefaultRegistries(): Registry[] {
  return [...defaultRegistries];
}

export async function getConfiguredRegistries(
  packageFile: string
): Promise<Registry[] | undefined> {
  // Valid file names taken from https://github.com/NuGet/NuGet.Client/blob/f64621487c0b454eda4b98af853bf4a528bef72a/src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs#L34
  const nuGetConfigFileNames = ['nuget.config', 'NuGet.config', 'NuGet.Config'];
  // normalize paths, otherwise startsWith can fail because of path delimitter mismatch
  const nuGetConfigPath = await findUpLocal(
    nuGetConfigFileNames,
    upath.dirname(packageFile)
  );
  if (!nuGetConfigPath) {
    return undefined;
  }

  logger.debug(`Found NuGet.config at ${nuGetConfigPath}`);
  const nuGetConfig = await readFileAsXmlDocument(nuGetConfigPath);
  if (!nuGetConfig) {
    return undefined;
  }

  const packageSources = nuGetConfig.childNamed('packageSources');
  if (!packageSources) {
    return undefined;
  }

  const registries = getDefaultRegistries();
  for (const child of packageSources.children) {
    if (child.type === 'element') {
      if (child.name === 'clear') {
        logger.debug(`clearing registry URLs`);
        registries.length = 0;
      } else if (child.name === 'add') {
        const isHttpUrl = regEx(/^https?:\/\//i).test(child.attr.value);
        if (isHttpUrl) {
          let registryUrl = child.attr.value;
          if (child.attr.protocolVersion) {
            registryUrl += `#protocolVersion=${child.attr.protocolVersion}`;
          }
          logger.debug(`Adding registry URL ${registryUrl}`);
          registries.push({
            name: child.attr.key,
            url: registryUrl,
          });
        } else {
          logger.debug(
            { registryUrl: child.attr.value },
            'ignoring local registry URL'
          );
        }
      }
      // child.name === 'remove' not supported
    }
  }
  return registries;
}
