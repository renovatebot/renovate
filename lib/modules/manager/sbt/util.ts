import { regEx } from '../../../util/regex';
import { get } from '../../versioning';
import * as mavenVersioning from '../../versioning/maven';

/*
  https://www.scala-sbt.org/release/docs/Cross-Build.html#Publishing+conventions
 */
export function normalizeScalaVersion(str: string): string {
  // istanbul ignore if
  if (!str) {
    return str;
  }
  const versioning = get(mavenVersioning.id);
  if (versioning.isVersion(str)) {
    // Do not normalize unstable versions
    if (!versioning.isStable(str)) {
      return str;
    }
    // Do not normalize versions prior to 2.10
    if (!versioning.isGreaterThan(str, '2.10.0')) {
      return str;
    }
  }
  if (regEx(/^\d+\.\d+\.\d+$/).test(str)) {
    return str.replace(regEx(/^(\d+)\.(\d+)\.\d+$/), '$1.$2');
  }
  // istanbul ignore next
  return str;
}
