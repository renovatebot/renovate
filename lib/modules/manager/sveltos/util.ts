import { regEx } from '../../../util/regex';

export function removeRepositoryName(
  repositoryName: string,
  chartName: string,
): string {
  const repoNameWithSlash = regEx(`^${repositoryName}/`, undefined, false);
  let modifiedChartName = chartName.replace(repoNameWithSlash, '');

  modifiedChartName = modifiedChartName.replace(/\/+$/, '');

  return modifiedChartName;
}
