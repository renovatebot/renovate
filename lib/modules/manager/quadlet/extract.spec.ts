import { codeBlock } from 'common-tags';
import { DockerDatasource } from '../../datasource/docker';
import type { ExtractConfig } from '../types';
import { extractPackageFile } from '.';
import { partial } from '~test/util';

const config = partial<ExtractConfig>({});

const configAliases = partial<ExtractConfig>({
  registryAliases: {
    'quay.io': 'registry.internal/mirror/quay.io',
  },
});

const packageFile = 'values.yaml';

describe('modules/manager/quadlet/extract', () => {
  describe('extractPackageFile()', () => {
    it('returns null for invalid quadlet file content', () => {
      const result = extractPackageFile('nothing here: [', packageFile, config);
      expect(result).toBeNull();
    });

    it('returns null for empty yaml file content', () => {
      const result = extractPackageFile('', packageFile, config);
      expect(result).toBeNull();
    });

    it('extracts from quadlet container unit', () => {
      const simple = codeBlock`
      [Container]
      Image=docker.io/library/alpine:3.22
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: '3.22',
            depName: 'docker.io/library/alpine',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('extracts from quadlet image unit', () => {
      const simple = codeBlock`
      [Image]
      Image=docker.io/library/alpine:3.22
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: '3.22',
            depName: 'docker.io/library/alpine',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('extracts from quadlet volume unit', () => {
      const simple = codeBlock`
      [Volume]
      Image=docker.io/library/alpine:3.22
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: '3.22',
            depName: 'docker.io/library/alpine',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('handles docker prefix', () => {
      const simple = codeBlock`
      [Volume]
      Image=docker://docker.io/library/alpine:3.22
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: '3.22',
            depName: 'docker.io/library/alpine',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('handles docker-daemon prefix', () => {
      const simple = codeBlock`
      [Volume]
      Image=docker-daemon:docker.io/library/alpine:3.22
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: '3.22',
            depName: 'docker.io/library/alpine',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('does not extract an image file reference', () => {
      const simple = codeBlock`
      [Container]
      Image=foo.image
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toBeNull();
    });

    it('does not extract an build file reference', () => {
      const simple = codeBlock`
      [Container]
      Image=foo.build
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toBeNull();
    });

    it('extract data from file with registry aliases', () => {
      const aliasFile = codeBlock`
      [Container]
      Image=quay.io/metallb/controller:v0.13.10
      `;

      const result = extractPackageFile(aliasFile, packageFile, configAliases);
      expect(result).toMatchObject({
        deps: [
          {
            currentValue: 'v0.13.10',
            depName: 'quay.io/metallb/controller',
            packageName: 'registry.internal/mirror/quay.io/metallb/controller',
            datasource: DockerDatasource.id,
          },
        ],
      });
    });

    it('handles an unsuccessful parse', () => {
      const simple = codeBlock`
      [Container]
      `;

      const result = extractPackageFile(simple, packageFile, config);
      expect(result).toBeNull();
    });
  });
});
