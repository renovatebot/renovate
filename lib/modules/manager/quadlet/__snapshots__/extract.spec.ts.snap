// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`modules/manager/quadlet/extract > extractPackageFile() > extracts from quadlet container file 1`] = `
{
  "deps": [
    {
      "currentDigest": undefined,
      "currentValue": "1.21",
      "datasource": "docker",
      "depName": "docker.io/alpine",
      "depType": "image",
      "packageName": "docker.io/alpine",
    },
  ],
}
`;

exports[`modules/manager/quadlet/extract > extractPackageFile() > extracts from quadlet container unit 1`] = `
{
  "deps": [
    {
      "currentDigest": undefined,
      "currentValue": "1.21",
      "datasource": "docker",
      "depName": "docker.io/alpine",
      "depType": "image",
      "packageName": "docker.io/alpine",
    },
  ],
}
`;

exports[`modules/manager/quadlet/extract > extractPackageFile() > extracts from quadlet image unit 1`] = `
{
  "deps": [
    {
      "currentDigest": undefined,
      "currentValue": "1.21",
      "datasource": "docker",
      "depName": "docker.io/alpine",
      "depType": "image",
      "packageName": "docker.io/alpine",
    },
  ],
}
`;

exports[`modules/manager/quadlet/extract > extractPackageFile() > extracts from quadlet volume unit 1`] = `
{
  "deps": [
    {
      "currentDigest": undefined,
      "currentValue": "1.21",
      "datasource": "docker",
      "depName": "docker.io/alpine",
      "depType": "image",
      "packageName": "docker.io/alpine",
    },
  ],
}
`;
